proposals/explicit-backing-fields.md

```diff
@@ -0,0 +1,322 @@
+
+```kotlin
+var someStrangeExample: Int
+    lateinit field: String
+    get() = field.length
+    set(value) {
+        field = value.toString()
+    }
+```
+
+#### Initializer
+
+If there is an explicit backing field, the property must not declare an initializer.
+
+If the explicit backing field is not `lateinit`, it must have an initializer. For `lateinit` properties initializers are forbidden.
+
+#### Mutability
+
+For now, we assume `val` properties have immutable backing fields, and `var` properties have mutable ones.
+
+#### Forbidden
+
+Explicit backing fields are not allowed inside interfaces or abstract properties.
+
+Explicit backing fields are not allowed for delegated properties.
+
+### Smart Type Narrowing
+
+#### Rules
+
+If:
+
+1. the compiler can guarantee the property getter returns the same instance as the one stored in the backing field
+2. the type of that instance is compatible with the property type
+3. the backing field visibility allows a hypothetical direct access
+4. there is no ambiguity in such an access
+
+it can then narrow down the returned instance type at the call site without loss of functionality. This is what is meant by the words _smart type narrowing_.
+
+The formal checks corresponding to the above rules are:
+
+1. the property does not have a custom getter, only the default one
+2. `A :> B`
+3. accessing:
+   1. a private backing field of a class member property within the class
+   2. a private backing field of a top-level property within the file
+   3. an internal backing field within the module
+4. property must be final, otherwise an overridden version would be able to provide a custom getter that may return a different instance
+
+#### Example
+
+```kotlin
+class MyClass {
+    val items: List<String>
+    	field = mutableListOf("a", "b")
+    
+    fun registerItem(item: String) {
+        items.add(item) // Viewed as MutableList<String>
+    }
+}
+
+fun test() {
+    val it = MyClass()
+    it.items // Viewed as a List<String>
+}
+```
+
+## Alternatives
+
+### Initial Proposal
+
+Initially, the following syntax was suggested:
+
+```kotlin
+private val items = mutableListOf<Item>()
+	public get(): List<Item>
+```
+
+> In above example `items` is `MutableList<Item>` when accessed privately inside class and read-only `List<String>` when accessed from outside class.
+
+In fact, the above syntax brings in an incorrect mental model: it says _'There is a private property `it` with some *part* that declares its public behavior'_.
+
+Attempt to add support for the above syntax led to multiple redundant complications (see the [problems section](https://github.com/matejdro/KEEP/blob/private_public_property_types/proposals/private_public_property_types.md#questions-to-consider) + unclear override mechanics).
+
+## Future Enhancements
+
+### Direct Backing Field Access
+
+It might be handy to allow the direct access to the property backing field via syntax like `myProperty#field`. 
```

Comment:
It is worth explicitly noting here that this is not the actually proposed syntax.